### PR TITLE
Ft create blog posts 153727855

### PR DESCRIPTION
**Pull request by LehruAngela**
## PR Status
**READY**

## Story ID
#153727855

## Added Migrations
YES

## Description
Users should be able to write blog posts on the health checks site.
A basic blog that allows logged in users to create categories, write blog posts and share blog posts.

## New Packages Added
django-pagedown

## Related PRs 
List related PRs against other branches:
None

## Todos
- [x] Tested and working locally
- [x] Tested and working on development environment
- [x] Added more Unit tests (if appropriate)
- [ ] Reviewed and approved by a TTL
- [ ] Deployment to team Heroku account

## Impacted Areas in the application
In hc/front, we changed the models, views, urls and forms files
In templates, we added blog_posts, readblog, create_blog html files
In hc/front/tests, we added a test_blogs file

## Known Issues (optional)
None

### CODE REVIEW

File: hc/accounts/tests/test_check_extended_timeout_period.py [hunk 1]
```diff
@@ -14,7 +14,15 @@ def test_it_works(self):
         
         period = 5184000
         url = "/checks/%s/timeout/" % self.check.code
+<<<<<<< HEAD
```
Comment (johnseremba) [inline hc/accounts/tests/test_check_extended_timeout_period.py]: Fix your merge conflicts properly

File: hc/front/forms.py [hunk 2]
```diff
@@ -19,11 +20,22 @@ def clean_tags(self):
 
 #increase max value in forms to allow upto 60days
 class TimeoutForm(forms.Form):
+<<<<<<< HEAD
```
Comment (johnseremba) [inline hc/front/forms.py]: fix conflicts

File: hc/front/models.py [hunk 3]
```diff
@@ -1,4 +1,30 @@
...
+
+    def __str__(self):
+        return f"{self.name}"
+
+    
+class Blog_post(models.Model):
+    title = models.CharField(max_length=100)
+    content = models.TextField()
+    published = models.DateTimeField(null=True, blank=True)
+    category = models.ForeignKey(Category, on_delete = models.CASCADE)
```
Comment (johnseremba) [inline hc/front/models.py]: I guess this implies that when a `Category` is deleted, the subsequent `Blog_post`s are deleted too. Is this the kind of behaviour that you require?
Look at this [post](https://stackoverflow.com/questions/38388423/what-does-on-delete-do-on-django-models) for a better explanation.

File: hc/front/models.py [hunk 4]
```diff
@@ -1,4 +1,30 @@
...
+    user = models.ForeignKey(User, blank=True, null=True, on_delete = models.CASCADE)
+
+    def __str__(self):
+        return f"{self.title}"
+
+class Comment(models.Model):
+    comment = models.TextField()
+    blog = models.ForeignKey(Blog_post, blank=True, null=True )
+    user = models.ForeignKey(User, blank=True, null=True, on_delete = models.CASCADE)
+    published = models.DateTimeField(null=True, blank=True)
```
Comment (johnseremba) [inline hc/front/models.py]: The date a comment is published should never be null. Provide a default value of the current time and date.
Use 
`models.DateTimeField(default=datetime.now, blank=True)`
or 
`models.DateTimeField(auto_now_add=True, blank=True)`

File: hc/front/tests/test_blogs.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+from django.utils import timezone
+
+class BlogTest(BaseTestCase):
+    def setUp(self):
+        super(BlogTest, self).setUp()
+        self.client.login(username="alice@example.org", password="password")
+        self.category = Category(name='Tech')
+        self.category.save()
+        self.blog = Blog_post(title='Python',content='This is a blog on python', 
+                             category = self.category, published=timezone.now())
```
Comment (johnseremba) [inline hc/front/tests/test_blogs.py]: Since the arguments proceed to the next line, sanitize your code by putting each kwarg on a separate line.

File: hc/front/tests/test_blogs.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        query_category = Category.objects.get(name="Design")
+        self.assertEqual('Design', query_category.name)
+        self.assertEqual(response.status_code, 200)
+        self.assertTemplateUsed(response, 'front/create_blog.html')
+
+
+    def test_create_blog(self):
+        """Test to check if blogs are created"""
+        url = reverse('hc-create_blog')
+        data = {'category_name': ['Tech'], 'title': ['Html'], 'content': ['This is a blog on html'], 'create_blog': ['']}
```
Comment (johnseremba) [inline hc/front/tests/test_blogs.py]: The line is too long. It is good practice to define dictionaries in a way that allows each Key Value pair to be on a separate line.
i.e;
```
data = { 
'category_name': ['Tech'], 
'title': ['Html'], 
'content': ['This is a blog on html'], 
'create_blog': ['']
}
```
That increases readability.

File: hc/front/views.py [hunk 7]
```diff
@@ -27,6 +32,107 @@ def pairwise(iterable):
     return zip(a, b)
 
 
+def blogs(request, filter_by):
+    num = 3
+    category = Category.objects.all()
+    if int(filter_by) > 0:
```
Comment (johnseremba) [inline hc/front/views.py]: Any thing less than zero is falsy, and the reverse is truthy.
so, just use;
`if filter_by:`

File: hc/front/views.py [hunk 8]
```diff
@@ -27,6 +32,107 @@ def pairwise(iterable):
     return zip(a, b)
 
 
+def blogs(request, filter_by):
+    num = 3
```
Comment (johnseremba) [inline hc/front/views.py]: Why do you hardcode `num = 3` you need to add a comment on why you did that.

File: hc/front/views.py [hunk 9]
```diff
@@ -27,6 +32,107 @@ def pairwise(iterable):
...
+    category = Category.objects.all()
+    if int(filter_by) > 0:
+        category_id = Category.objects.get(pk=filter_by).id
+        stories = Blog_post.objects.filter(category=category_id)
+    else:
+        stories = Blog_post.objects.all()
+    list_of_stories = [story for story in stories]
+    blogs = [list_of_stories[i:i+num] for i in range(0, len(list_of_stories), num)]
```
Comment (johnseremba) [inline hc/front/views.py]: avoid single letter variable names such as `i`. 
I must be saying this for the 1000th time.

Comment (johnseremba) [inline hc/front/views.py]: Nice comprehension here. However, add a comment about what you want to achieve for code easy code maintainability.

File: hc/front/views.py [hunk 10]
```diff
@@ -27,6 +32,107 @@ def pairwise(iterable):
...
+            title = request.POST['title'] 
+            blog = form.cleaned_data['content']
+            selected_category = request.POST['category_name']
+            category = Category.objects.get(name=selected_category)
+            published = timezone.now()
+            user = request.user
+            blog = Blog_post(title=title, content=blog, category=category,
+                             published=published, user=user)
+            blog.save()
+            messages.add_message(request, messages.INFO, 'Successfully created blog')
```
Comment (johnseremba) [inline hc/front/views.py]: Be consistent with the quotation marks you choose to use. either go with double or single.

File: hc/front/views.py [hunk 11]
```diff
@@ -27,6 +32,107 @@ def pairwise(iterable):
...
+
+
+def read_blog(request, pk):
+    comment_form = CreateCommentForm(request.POST)
+    blog = Blog_post.objects.get(pk=pk)
+    featured = Blog_post.objects.get(pk=pk)
+    comments = Comment.objects.filter(blog = blog.id)
+    url = f"http://localhost:8000/blog/read_blog/{pk}"
+    ctx = {
+        "blog": blog,
```
Comment (johnseremba) [inline hc/front/views.py]: be consistent with your quotation marks

File: hc/front/views.py [hunk 12]
```diff
@@ -27,6 +32,107 @@ def pairwise(iterable):
...
+          }
+    return render(request, "front/create_blog.html", ctx)
+
+
+def read_blog(request, pk):
+    comment_form = CreateCommentForm(request.POST)
+    blog = Blog_post.objects.get(pk=pk)
+    featured = Blog_post.objects.get(pk=pk)
+    comments = Comment.objects.filter(blog = blog.id)
+    url = f"http://localhost:8000/blog/read_blog/{pk}"
```
Comment (johnseremba) [inline hc/front/views.py]: Use the `str.format()` function instead.

File: static/css/blog.css [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+
+>>>>>>> 1b6c6c7389d654a415c25e22d23119f2a605fafa
+   .blog-item img {
+     width: 65%;
+     overflow: hidden;
+     padding: 0;
+     display: block;
+     margin-left: auto;
+     margin-right: auto;
+   }
```
Comment (johnseremba) [inline static/css/blog.css:43]: FIX YOUR MERGE CONFLICTS

File: templates/front/my_checks.html [hunk 14]
```diff
@@ -116,11 +128,19 @@ <h4 class="update-timeout-title">Name and Tags</h4>
...
+                            <span>Grace Time</span> When a check is late, how much time to wait until alert is sent.
+                        </p>
+                        <p>
+=======
                         </p>
                         <p>
                             <span>Grace Time</span> When a check is late, how much time to wait until alert is sent.
                         </p>
                         <p>
+>>>>>>> Nag interval Build Error Fix #153727845 (#19)
```
Comment (johnseremba) [inline templates/front/my_checks.html]: FIX YOUR MERGE CONFLICTS

File: templates/front/my_checks_desktop.html [hunk 15]
```diff
@@ -65,11 +65,16 @@
                         <span class="checks-subline">
                         {{ check.grace|hc_duration }}
                         </span>
+<<<<<<< HEAD
```
Comment (johnseremba) [inline templates/front/my_checks_desktop.html]: SAME AS ABOVE

File: templates/front/my_checks_desktop.html [hunk 16]
```diff
@@ -217,4 +345,71 @@
         </div>
     </div>
 
-</div>
\ No newline at end of file
+</div>
+=======
```
Comment (johnseremba) [inline templates/front/my_checks_desktop.html]: SAME AS ABOVE

File: templates/front/readblog.html [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+        </div>
+</main>
+
+<script
+  src="http://code.jquery.com/jquery-1.12.4.min.js"
+  integrity="sha256-ZosEbRLbNQzLpnKIkEdrPv7lOy9C27hHQ+Xp8a4MxAQ="
+  crossorigin="anonymous"></script>
+<script>
+
+    $(function(){
```
Comment (johnseremba) [inline templates/front/readblog.html]: Javascript code should never be in the HTML template file.
Create another `.js` file for you `js` code.

**Review by johnseremba — CHANGES_REQUESTED** [review]
Urgently work on the requested changes.

File: hc/front/models.py (hunk 4 above)
Comment (LehruAngela) [reply hc/front/models.py]: Noted and worked on

File: hc/front/tests/test_blogs.py (hunk 5 above)
Comment (LehruAngela) [reply hc/front/tests/test_blogs.py]: Noted and worked on

File: hc/front/tests/test_blogs.py (hunk 6 above)
Comment (LehruAngela) [reply hc/front/tests/test_blogs.py]: Noted and worked on

File: hc/front/views.py (hunk 8 above)
Comment (LehruAngela) [reply hc/front/views.py]: Noted and added a comment clarifying its purpose

File: hc/front/views.py (hunk 7 above)
Comment (LehruAngela) [reply hc/front/views.py]: Noted and worked on

File: hc/front/views.py (hunk 9 above)
Comment (LehruAngela) [reply hc/front/views.py]: We've changed it

File: static/css/blog.css (hunk 13 above)
Comment (LehruAngela) [reply static/css/blog.css:43]: Noted and fixed

File: templates/front/my_checks.html (hunk 14 above)
Comment (LehruAngela) [reply templates/front/my_checks.html]: Noted and fixed

File: templates/front/my_checks.html [hunk 18]
```diff
@@ -116,11 +116,19 @@ <h4 class="update-timeout-title">Name and Tags</h4>
                     <div class="update-timeout-terms">
                         <p>
                             <span>Period</span> Expected time between pings.
+<<<<<<< HEAD
```
Comment (johnseremba) [inline templates/front/my_checks.html]: Fix this conflict too

Comment (LehruAngela) [reply templates/front/my_checks.html]: Fixed

File: hc/front/views.py (hunk 12 above)
Comment (LehruAngela) [reply hc/front/views.py]: Noted and fixed

File: hc/front/tests/test_blogs.py [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+        response = self.client.get(reverse('hc-blog', kwargs={'filter_by':0}))
+        self.assertEqual(response.status_code, 200)
+        self.assertTemplateUsed(response, 'front/blog_posts.html')
+
+
+    def test_edit_blog(self):
+            """Test to check if blogs are edited"""
+            blog = Blog_post.objects.get(title="Python")
+            data = { 'title': ['ben'],
+            'content': ['If it were onlya question of qualitycv'], 'create_blog': ['']}
```
Comment (johnseremba) [inline hc/front/tests/test_blogs.py]: Beware of typographic errors!

File: hc/front/tests/test_blogs.py [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+            response = self.client.post(reverse('hc-edit_blog', kwargs={'pk':blog.id}), data)
+            edited_blog = Blog_post.objects.filter(pk=blog.id).first()
+            self.assertEqual(response.status_code, 302)
+            self.assertEqual('ben', edited_blog.title)
```
Comment (johnseremba) [inline hc/front/tests/test_blogs.py]: you should use `2` blank lines after a function, not `3`

File: hc/front/tests/test_blogs.py [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+        blog = Blog_post.objects.get(title="Python")
+        response = self.client.get(reverse('hc-delete_blog', kwargs={'pk':blog.id}))
+        deleted_blog = Blog_post.objects.filter(pk=blog.id).first()
+        self.assertEqual(response.status_code, 302)
+        self.assertNotEqual(blog, deleted_blog)
+
+
+   
+        
+
```
Comment (johnseremba) [inline hc/front/tests/test_blogs.py]: Unnecessary whitespace.

File: hc/front/views.py [hunk 22]
```diff
@@ -27,6 +32,112 @@ def pairwise(iterable):
...
+    }
+    return render(request, 'front/blog_posts.html', ctx)
+
+
+def create_blog(request):
+    """function to create a category and a blog """
+    form = CreateBlogPost(request.POST)
+    category_form = CreateCategory(request.POST)
+    if request.method == 'POST':
+        if 'new_category' in request.POST:
```
Comment (johnseremba) [inline hc/front/views.py]: Unnecessary nesting of `if` statements. Since the inner `if` has no else and it is the only one, you can conjoin that logic to the outer conditional statement. Something like this;

```
if 'new_category' in request.POST and 'category_form.is_valid():
  # your logic goes here
```

File: hc/front/views.py [hunk 23]
```diff
@@ -27,6 +32,112 @@ def pairwise(iterable):
...
+                published = timezone.now()
+                user = request.user
+                blog = Blog_post(title=title, content=blog, category=category,
+                                published=published, user=user)
+                blog.save()
+                messages.add_message(request, messages.INFO, 'Successfully created blog')
+                return redirect(read_blog, pk=blog.id)
+    else:
+        categories = Category.objects.all()
+        ctx = {
```
Comment (johnseremba) [inline hc/front/views.py:76]: For neat code, add a space between the `key` and `value`. Did you follow PEP8 coding standards?
```
'category': categories,
'form': CreateBlogPost({'content':'Write Blog'}),
'category_form': category_form,
 'edit': False
```

File: hc/front/views.py [hunk 24]
```diff
@@ -27,6 +32,112 @@ def pairwise(iterable):
...
+        category_form = CreateCategory(request.POST)
+        if "create_blog" in request.POST:
+            if form.is_valid():
+                blog.category = blog.category
+                blog.title = request.POST['title']
+                blog.content = request.POST['content']
+                blog.publish = timezone.now()
+                blog.save()
+                messages.add_message(request, messages.INFO, 'Successfully edited blog')
+                return redirect(read_blog, pk)
```
Comment (johnseremba) [inline hc/front/views.py:126]: So many nested `if`s. strive not to have more than 2.
conjoin `form.is_valid()` with the outer conditional statement.

**Review by johnseremba — CHANGES_REQUESTED** [review]

File: hc/front/views.py (hunk 24 above)
Comment (LehruAngela) [reply hc/front/views.py:126]: Noted and fixed

File: hc/front/views.py (hunk 23 above)
Comment (LehruAngela) [reply hc/front/views.py:76]: Noted and fixed

**Review by johnseremba — APPROVED** [review]
